### DISCUSSION TITLE
Errors when running PyTest with FastAPI Users & Beanie

### DISCUSSION THREAD

**Discussion by ghost** [fastapi-users/fastapi-users#1203] [Q&A]
Hi All,

My unit tests are failing when I call Beanie's Document.fetch_all_links() method (from within one of my endpoints in the app - which works when I'm not testing) via a fixture in conftest.py:

```
src/wj/location/service.py:43: in create
    await created_location.fetch_all_links()
/usr/local/lib/python3.11/site-packages/beanie/odm/documents.py:952: in fetch_all_links
    await asyncio.gather(*coros)
/usr/local/lib/python3.11/site-packages/beanie/odm/documents.py:943: in fetch_link
    values = await Link.fetch_list(ref_obj, fetch_links=True)
/usr/local/lib/python3.11/site-packages/beanie/odm/fields.py:175: in fetch_list
    model_class = link.model_class
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def __getattribute__(self, item):
>       res = super(LazyModel, self).__getattribute__(item)
E       AttributeError: 'User' object has no attribute 'model_class'

/usr/local/lib/python3.11/site-packages/lazy_model/main.py:117: AttributeError
```

So I'm leaning towards thinking that I'm not initializing FastAPI-Users properly when running tests?

I'm going to keep digging - but if anyone has an example of a working PyTest + Fastapi Users + beanie setup I'd greatly appreciate it!

**Comment by frankie567** [comment]
Did you make sure to call [init_beanie](https://beanie-odm.dev/tutorial/initialization/) before starting the tests?

**Reply by ghost** [reply]
Here's where I get confused - and thanks so much for writing back. in my main.py I've got the standard @app.on_event("startup") method wherein I call init_beanie. Am I to understand that I should be doing that separately in conftest.py? Do I do a full API startup in conftest.py?

**Reply by ghost** [reply]
I added some logging, and when running my tests I see that yes, beanie is being initialized - and I can create some objects and store them in the DB.

```
2023-05-04 20:39:00,169 - INFO - Initializing Beanie
2023-05-04 20:39:00,180 - INFO - ODM initialization complete
2023-05-04 20:39:00,448 - INFO - Created default user: admin@example.com
2023-05-04 20:39:00,449 - INFO - Creating company: Big Bird's Seed Emporium
2023-05-04 20:39:00,455 - DEBUG - Fetching and storing coordinates for model Company : 645417e44de2960c774ee619
```

**Reply by frankie567** [reply]
Right, it seems correctly initialized.

How do you authenticate users in your tests? Are you overriding the `current_user` dependency with an arbitrary `User` object?

**Reply by ghost** [reply]
I've isolated it further - feels like it 's more of a Beanie problem... 

Previously I was:

1. creating an object, with a link to another
2. running new_object.fetch_all_links()

And that was failing. 

As a test, and this is not what I want to really be doing, but it works - I can:
1. create an object that has a link to another object
2. use the id of the newly created object to fetch itself from the DB, using fetch_links=True in the .get() call.

I'll see if I can re-create this in a small example and let Roman @ Beanie know.
